File: Alternate positive and negative numbers.java
class Solution {
    void rearrange(int arr[], int n) {
        // code here
        boolean needNeg=false;
        for(int i=0;i<n;i++){
            if(arr[i]>=0){
                //need pos
                if(needNeg==false){
                    needNeg=true;
                }else{
                    //find neg
                    int j=i+1;
                    while(j<n && arr[j]>=0){
                        j++;
                    }
                    if(j==n)return;
                    int ele=arr[j];
                    j--;
                    while(j>=i){
                        arr[j+1]=arr[j];
                        j--;
                    }
                    arr[i]=ele;
                    needNeg=false;
                }
            }else{
                //needed neg
                if(needNeg==true){
                    needNeg=false;
                }else{
                    //find pos
                    int j=i+1;
                    while(j<n && arr[j]<0){
                        j++;
                    }
                    if(j==n)return;
                    int ele=arr[j];
                    j--;
                    while(j>=i){
                        arr[j+1]=arr[j];
                        j--;
                    }
                    arr[i]=ele;
                    needNeg=true;
                }
            }
        }
    }
}

File: Alternate positive and negative numbers.cpp
class Solution{
public:

void rearrange(int arr[], int n) {
	// code here
	stack<int> st,st1;
       for(int i=n-1;i>=0;i--){
           if(arr[i]>=0) st.push(arr[i]);
           else st1.push(arr[i]);
       }
       for(int i=0;i<n;i++){
           if(i%2==0 && !st.empty()){
               arr[i]=st.top();
               st.pop();
           }
           else if(i%2==1 && !st1.empty()){
               arr[i]=st1.top();
               st1.pop();
           }else if(!st.empty()){
               arr[i]=st.top();
               st.pop();
           }else{
               arr[i]=st1.top();
               st1.pop();
           }
       }
   }
};